src/transformers/models/llama/tokenization_llama.py

```diff
@@ -0,0 +1,209 @@
+# coding=utf-8
+# Copyright 2022 The FAIR team of Meta AI and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tokenization classes for LLaMA."""
+import os
+import re
+from shutil import copyfile
+from typing import Any, Dict, List, Optional, Tuple
+
+import sentencepiece as spm
+
+from ...tokenization_utils import PreTrainedTokenizer
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+VOCAB_FILES_NAMES = {"vocab_file": "tokenizer.model"}
+
+PRETRAINED_VOCAB_FILES_MAP = {}
+
+
+class LLaMATokenizer(PreTrainedTokenizer):
+    """
+    Construct a LLaMA tokenizer. Based on byte-level Byte-Pair-Encoding.
+
+    Args:
+        vocab_file (`str`):
+            Path to the vocabulary file.
+    """
+
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    model_input_names = ["input_ids", "attention_mask"]
+
+    def __init__(
+        self,
+        vocab_file,
+        unk_token="",
+        bos_token="",
+        eos_token="",
+        sp_model_kwargs: Optional[Dict[str, Any]] = None,
+        add_bos_token=False,
+        add_eos_token=False,
```

Comment:
Maybe change the default to `add_bos_token=True`?

Reply:
I'm actually not sure what the desired/default behavior for this should be. From the above discussion it looks like `add_eos_token=False` is a good idea, but I'm not sure in other HF pipelines (or other code-bases) whether we expect BOS to be added. For instance, GPT-2's tokenizer doesn't add special tokens by default.

Reply:
I would say that given that LLaMa by default adds `bos_token`. It was probably trained like that, so I would add this by default.

Reply:
But adding that bos_token leads to bad results, both in theory and practice (there are several users reporting this).

Reply:
But adding that EOS leads to bad results, both in theory and practice. In fact, there are several users reporting this (including me), and would have to resort to special ugly hacks to remove that EOS as in https://github.com/zsc/llama_infer/blob/d6e2b02a667d41ee9237933560fe1e89eef0ab0d/test_llam.py#L71

Reply:
Yeah this is expected, you never want to add <eos> at the end of your prompt, I think eos was used for training:
```
<bos>{TEXT}<eos>
```
The issue is when you generate, you want to feed a partial test and allow the model to keep on generating, this is usually done like this:
```
<bos>{PROMPT}
```
And then let the model generate until it find <eos> or fills max token.